### ISSUE TITLE
Question about Training Data Preparation for Sesame Models

### ISSUE THREAD

**Issue by thomaslwang** [thomasgauthier/csm-hf#2]
Issue Description
I have a question regarding how to properly prepare training data for Sesame models. According to the documentation:

Each model was trained with a 2048 sequence length (~2 minutes of audio) over five epochs.

I'm trying to understand how this translates to the JSONL data format we're using. My specific questions are:

Does each line in the JSONL file need to contain 2048 audio frames of continuous conversation?
Or can each line be a single conversation exchange, with the model automatically concatenating samples to reach the 2048 frame sequence length during training?

{
  "messages": [
    {
      "role": "speaker_0",
      "content": [
        {"type": "text", "text": "We have a chance for a new life here."},
        {"type": "audio", "url": "clips/example_audio.wav"}
      ]
    },
    {
      "role": "speaker_1",
      "content": [
        {"type": "text", "text": "Uncle?"},
        {"type": "audio", "url": "clips/response_audio.wav"}
      ]
    }
  ],
  "training_mask": [false, true]
}

**Comment by thomasgauthier** [comment]
So right now each jsonl line corresponds to one sequence or conversation (can contain many utterances). If too short, it is padded to length 2048. I believe what you are describing with concatenating multiple sequences to max out the 2048 context length is called sample packing, it requires proper handling of attention mask to make sure individual sequences are not attending to unrelated sequences. This is not supported right now but I will add it to the todo.

**Comment by thomaslwang** [comment]
Thank you for your quick response - that clarifies my understanding about how each JSONL line corresponds to a sequence that's padded to 2048 if needed.
I have a follow-up question: I believe the current training approach you've described is for fine-tuning, correct? The 2048 sequence length mentioned in the Sesame documentation would be from the pre-training phase. Is there any support for pre-training in the current codebase, or are there plans to add this capability?
I'm interested in contributing to this open-source project and would be happy to collaborate with you on implementing these features. I have access to GPU resources for training and a dataset of approximately 1 million hours of English speech data that could be valuable for pre-training efforts.

**Comment by thomasgauthier** [comment]
Yes the main use case I had in mind was finetuning. Continual pretraining is supported in theory, just need to use a high batch size and low learning rate. But typically pretraining is done with highly optimized training code for multi-node setups. This repo was created with single GPU use case in mind so it's not at all optimized for pretraining. Using csm-hf to train on a million hours would realistically take months or even years.

However, the CSM 1B model released by Sesame has already been trained on a million hours of English and is highly capable as-is, I don't really think doing more pretraining would be worth it if not for learning a new language. If output quality is your ultimate goal, leveraging compute for large-scale reinforcement learning would give you more bang for your buck in my opinion.

Another option for you is to train a Orpheus model. Because it uses a simpler Llama architecture you could use any pretraining toolkit like Megatron without needing custom code.

**Comment by thomaslwang** [comment]
Thanks a lot for the response.
